### PR TITLE
feat: Migrate graphite parser to new style

### LINKED ISSUE #11356: Migrate graphite to new parser structure
More info can be found [here](https://github.com/influxdata/telegraf/issues/10482)

### PR DESCRIPTION
**Pull request by srebhan**
- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #11356  

This PR migrates the `graphite` parser to the new-style parser structure. The change is backward compatible. Furthermore, we cleanup the parser a bit.

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -491,6 +492,10 @@ func TestConfig_ParserInterfaceNewFormat(t *testing.T) {
 	require.Len(t, actual, len(formats))
 
 	for i, format := range formats {
+		if format == "graphite" {
+			fmt.Printf("actual=%v\n", actual[i])
+			fmt.Printf("expected=%v\n", expected[i])
+		}
```
Comment (powersj) [inline config/config_test.go:498]: is this left over debugging?

Comment (srebhan) [reply config/config_test.go]: Good catch! Fixed.

File: plugins/parsers/graphite/parser_test.go [hunk 2]
```diff
@@ -122,34 +122,25 @@ func TestTemplateApply(t *testing.T) {
...
-			t.Fatalf("err does not match.  expected %v, got %v", test.err, err)
-		}
-		if err != nil {
-			// If we erred out,it was intended and the following tests won't work
+		if test.err == "" {
+			require.NoError(t, err)
+		} else {
+			require.EqualError(t, err, test.err)
 			continue
 		}
```
Comment (sspaink) [inline plugins/parsers/graphite/parser_test.go:130]:
```suggestion
		if test.err != "" {
			require.EqualError(t, err, test.err)
			continue
		}
		require.NoError(t, err)
```

File: plugins/parsers/graphite/parser_test.go [hunk 3]
```diff
@@ -281,19 +272,17 @@ func TestParseLine(t *testing.T) {
...
```
Comment (sspaink) [inline plugins/parsers/graphite/parser_test.go:284]:
```suggestion
		if test.err != "" {
			require.EqualError(t, err, test.err)
			continue
		}
		require.NoError(t, err)
```

File: plugins/parsers/graphite/parser_test.go [hunk 4]
```diff
@@ -396,19 +385,17 @@ func TestParse(t *testing.T) {
...
-			t.Fatalf("err does not match.  expected [%v], got [%v]", test.err, err)
-		}
-		if err != nil {
-			// If we erred out,it was intended and the following tests won't work
+		if test.err == "" {
+			require.NoError(t, err)
+		} else {
+			require.EqualError(t, err, test.err)
 			continue
 		}
```
Comment (sspaink) [inline plugins/parsers/graphite/parser_test.go:397]:
```suggestion
		if test.err != ""  {
			require.EqualError(t, err, test.err)
			continue
		}
		require.NoError(t, err)
```

**Review by sspaink — APPROVED** [review]
Looks good! added some suggested changes but not required, just makes the code look a little nicer.

**Comment by srebhan** [comment]
Thanks @sspaink for the suggestions! I like them.

**Pull request merged**